mempool/clist_mempool.go

```diff
@@ -293,7 +293,9 @@ func (mem *CListMempool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo Tx
 
 	reqRes, err := mem.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{Tx: tx})
 	if err != nil {
-		mem.cache.Remove(tx)
+		if !mem.config.CacheKeepCheckTxInvalid {
+			mem.cache.Remove(tx)
+		}
 		return err
 	}
```

Comment:
This can be a bit dangerous if the ABCI app isn't in-process. If we're using e.g. a UNIX socket or gRPC for ABCI, and the application is temporarily unavailable such that this errors, we're going to permanently ban transactions that are entirely valid. We should ideally differentiate between IO errors and ABCI errors here, but I'm not sure if our current error types allow for that or not.

Reply:
Hm, or maybe I'm wrong, this might just retry it later -- I'm not very familiar with the mempool. It's a failure mode worth considering though.

Reply:
Right. However take into account that this options is disabled by default.